// @sap-cloud-sdk/odata-common/dist/request-builder/action-function-import-request-builder-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionFunctionImportRequestBuilderBase = void 0;
const request_builder_base_1 = require("./request-builder-base");
/**
 * Create OData request to execute a action or function import.
 * @typeparam ReturnT - Type of the function import return value.
 * @internal
 */
class ActionFunctionImportRequestBuilderBase extends request_builder_base_1.MethodRequestBuilder {
    /**
     * Base class for function  and actions imports.
     * @param responseTransformer - Transformation function for the response.
     * @param requestConfig - Request config for a action or function import.
     */
    constructor(responseTransformer, requestConfig) {
        super(requestConfig);
        this.responseTransformer = responseTransformer;
    }
    /**
     * Execute request.
     * @param destination - Destination or DestinationFetchOptions to execute the request against.
     * @param dataAccessor - A function to modify the received response data.
     * @returns A promise resolving to the requested return type.
     */
    async execute(destination, dataAccessor) {
        return this.executeRaw(destination).then(response => {
            const data = dataAccessor
                ? { d: dataAccessor(response.data) }
                : response.data;
            return this.responseTransformer(data);
        });
    }
    /**
     * Execute request and return an [[HttpResponse]].
     * @param destination - Destination or DestinationFetchOptions to execute the request against.
     * @returns A promise resolving to an [[HttpResponse]].
     */
    async executeRaw(destination) {
        return this.build(destination).then(request => request.execute());
    }
}
exports.ActionFunctionImportRequestBuilderBase = ActionFunctionImportRequestBuilderBase;
